### DISCUSSION TITLE
How to not execute code with `quarto render`

### DISCUSSION THREAD

**Discussion by danmackinlay** [quarto-dev/quarto-cli#10984] [Q&A]
### Description

According to the CLI help, the following will not execute code in the `.qmd` files in my project

```bash
quarto render --no-execute 
```

However, for me, this seems to **force** execution.
If I wish to not execute the files I need to run 

```bash
quarto render --cache --user-freezer
```

This leads me to suspect that there is something I do not understand about the `--execute`/`--no-execute` flags. Does anyone have insight?

**Comment by mcanouil** [comment]
I believe `--(no-)execute` flag are for Jupyter engines.

**Comment by cderv** [comment]
@danmackinlay Can you share more on the context ? 

- Are you working with computation in Jupyter, Julia or Knitr ? 
- Which type of project are you working on ? 

Some context on execution 

- When working with `.ipynb` file, no execution will happen by default
	- `--execute` and `--no-execute` flags can help control (passed internally to `nbconvert`). 
- When working with `.qmd`, files need to execute to have results output in them so computation is executed by default. 

Execution of cells can also be controlled with [execution option](https://quarto.org/docs/computations/execution-options.html#output-options)
````yaml
execute:
  eval: false
````

When you have results computed once, and you would like to gain speed for next rendering, and not recomputing, two mechanisms can be leveraged: 

- Cache which is engine dependent. Cells are executed, but caching mechanism applies at the engine execution level. `--cache` will activate this for engine globally. It can also be controlled more closely to the engine. 
- Freeze which is Quarto specific. In that case, computation does not run at all. Quarto skips this step by loading a stored state of markdown content and results to recreate the intermediate .md file to be used in conversion. `--use-freezer` forces the use of the cache - which can be useful for an incremental render. 

Things to know about freeze:
- on a project with `freeze: true` or `freeze: auto` set  
	- `quarto render` will always use the freezer if it exists. 
	- `quarto render <a file in the project>` will not use the freezer, unless `--use-freezer` is set.  

- If freezer is used, then caching does not matters. 

I hope this helps understand the logic, and helps you pinpoint what is not clear so that you can ask more precisely and share example. Thank you!

**Reply by cderv** [reply]
To be more precise on my previous answers. 

`--execute` and `--no-execute` controls execution of the engine. This is true for Jupyter, and for Knitr too. In **knitr** context, it will be equivalent to `eval: false` I mentioned too. 

However, this means that the computation dependency needs to be present !! The code cell won't be evaluated but it still means R and **knitr** (+ **rmarkdown**) needs to be available and found by Quarto which will pass the information of `eval: false`. 

If you want to build with Quarto on an environment with no computation available, you need to render everything first and freeze the output, commit the frozen state and re-render with `quarto render` using this freezer. That is the only way: https://quarto.org/docs/projects/code-execution.html#freeze

**Reply by danmackinlay** [reply]
Interesting! I have frozen outputs and they are committed into version control; but for me the `--no-execute` flag causes quarto to ignore the frozen computation and execute the files. `--execute` does not, and no command line option does not

```bash
$ quarto render --use-freezer
[   1/1488] contact.qmd
[   2/1488] post/ismir2019.qmd
[   3/1488] post/regression_intro_2022.qmd
[   4/1488] post/age_equity_dating.qmd
[   5/1488] post/morandini_and_blamey.qmd
[   6/1488] post/masters_thesis.qmd
[   7/1488] post/llm_hacks_2024.qmd
[   8/1488] post/sxsw_2010.qmd
[   9/1488] post/jakarta_biennale_2009_realtime.qmd
^C⏎
$ quarto render --use-freezer --execute
[   1/1488] contact.qmd
[   2/1488] post/ismir2019.qmd
[   3/1488] post/regression_intro_2022.qmd
[   4/1488] post/age_equity_dating.qmd
[   5/1488] post/morandini_and_blamey.qmd   
^C⏎                                                                  
$ quarto render --use-freezer --no-execute
[   1/1488] contact.qmd
[   2/1488] post/ismir2019.qmd
[   3/1488] post/regression_intro_2022.qmd
[   4/1488] post/age_equity_dating.qmd

NOTE: Dependency discovery took 24 seconds during snapshot.
[… R code execution]
```

Are we sure that the `--no-execute` flag is not "flipped" in meaning? Except that I would not be the first person to notice it if this were the case. Tested on quarto 1.5.57 and 1.5.52.

**Reply by cderv** [reply]
I think this is from this line
https://github.com/quarto-dev/quarto-cli/blob/beca9e2c335ae5cbfa3e6d90d14093cbf49cadb3/src/command/render/render-files.ts#L161-L164

When `--no-execute` is set, the project and files will be considered as having no execution and so freezing is not possible. 

Maybe this is not correct way to do it. 

What are you expecting with `quarto render --use-freezer --no-execute` and not just `quarto render --use-freezer` which should be the default when doing `quarto render` ? 

`--execute` and `--no-execute` are ipynb related really, and this logic was added in 56cfe322767d417c0c54123e4e637fef4957a87c so to change it we need to understand the different use case really well

**Reply by danmackinlay** [reply]
>What are you expecting with `quarto render --use-freezer --no-execute` and not just `quarto render --use-freezer` which should be the default when doing quarto render ?

I don't have an opinion about what the behaviour should be in the case of this particular combination of flags, _per se_. This is just one of the CLI invocations I tried while debugging. 

For me, no matter what other flag I combine it with, [`--no-execute` always causes code to execute, and leaving `--no-execute` off allows the code to not execute](https://github.com/quarto-dev/quarto-cli/discussions/10984#discussioncomment-10866632). _This_ behaviour is the unintuitive one, not how it interacts with the freeze system.

I think for now I will just resign myself to the fact that the flags do the opposite of what I as a user would naively assume; I don't need to no the reason if this makes some kind of deeper sense to someone else, as long as it is predictable.

**Reply by cderv** [reply]
> For me, no matter what other flag I combine it with, https://github.com/quarto-dev/quarto-cli/discussions/10984#discussioncomment-10866632. This behaviour is the unintuitive one, not how it interacts with the freeze system.

To help me see clearer, what do you expect `--no-execute` to do ? When you set this flag, and you say it causes code to execute, to you mean your code cells are ran ? Or just do you expect that no computation dependencies is needed ? 

I think the confusion may lie here
- `--no-execute` is a way to set at command line that cells shouln't be executed inside the document, but it does not imply no execution engine. 
- `freeze: true` or `freeze: auto` will allow to use `quarto render` without computations engine needed in the environment

I can see how this could be confusing. 🤔 I don't know how we can change this without breaking anything. At least we can do a documentation update to make that clear. 

Putting back here current state so that it is documented here:

* `--no-execute` will apply on .ipynb source file to ensure that cells are not re-executed on the notebook. It does not mean Jupyter is not needed here. 
* `--no-execute` on a `.qmd` file is equivalent to `eval: false`. Engine is still ran to produce the intermediate ..md file require for conversion 
* Using `freeze` is what really allows to remove the need of computation in the environment for next `quarto render` as the intermediate state (after computation by engine) is saved and reuse.

**Reply by mcanouil** [reply]
Quarto CLI execute flags are not described (in `quarto render --help`) to be Jupyter specific last time I checked.
We know this, but users cannot.

**Reply by cderv** [reply]
They are not `.ipynb` specific. They work also for `.qmd`. They just mean `#| eval: false`. 

Difference between `.ipynb` and `.qmd` is that
* `.ipynb` could have been executed elsewhere (not with quarto), and this format stores results. So when you do `quarto render` we re-use those result by default. `--no-execute` is the default. 
* `.qmd` is a format which don't store result. So default is to execute by default the cells. `--no-execute` allows to not execute the cells (`eval: false`) which is different than not running engine at all. Engine needs to be ran for a .qmd to be rendered. 

In short, we could say `--no-execute` does not mean `engine: markdown`. It means: Render the document without executing the computations cells. But only `freeze` can remove the need of an engine because document have been rendered at least one with an engine. 

Hope it clarifies

**Reply by mcanouil** [reply]
To summarise briefly:
- `--execute` (`--no-execute`) force evaluation (or the opposite) of all code cells.
- `--use-freezer` => reuse computed code cells (thus "don't evaluate again")

---

I think the CLI help could made a bit clearer on that, as currently it is:
- `--execute`: "Execute code (--no-execute to skip execution)"
- `--use-freezer`: "Force use of frozen computations for incremental file render"

Maybe something like:
- `--execute`: Force execution of all code cells (`eval: true`). --no-execute to skip all code cells (`eval: false`).
- `--use-freezer`: Force (re)use of frozen code cells computations (requires `freeze` to be enabled).

**Reply by cderv** [reply]
>  think the CLI help could made a bit clearer on that

On a documentation improvement request where we can discuss the wording. I am not completely ok with all of it (`use-freezer` only apply to incremental render as current arg doc says which your proposal is removing). 

Probably a blog post explaining all this (how to control computation in quarto) could be more useful to document all this than just updating the help. 🤷

**Reply by mcanouil** [reply]
The definition of "incremental" is tricky though.
What does it mean in a "default" project (and/or in a single file project)?
I'm not sure users know that by incremental Quarto means `quarto render full/path/to/file.qmd` and not `quarto render`.

So indeed, a small blog post about workflow with freeze/cache/execute could be a good subject to discuss.

**Reply by cderv** [reply]
Here is the definition for incremental rendering in the project section: https://quarto.org/docs/projects/code-execution.html#incremental-render

If not clear enough, you can submit enhancement proposal for the documentation.

**Reply by mcanouil** [reply]
I've never seen this part.^^
Thanks!

I would not have expected or searched for "Incremental Render" in "Managing Execution". It makes sense and is consistent with Quarto CLI help though.

**Reply by cderv** [reply]
Reading a whole manual once is sometimes the only way (and quite useful) to know everything that is possible! 😉

**Reply by mcanouil** [reply]
You forget about the regular updates ;)

**Comment by danmackinlay** [comment]
Fascinating! OK thanks @cderv that was an excellent explanation.

My case is that I have a site with `.qmd` files, mostly executing R code.

There is a phenomenon which I find hard to reconcile with the cache flow logic you describe, so if you have a moment to exaplin it further I woudl be grateful.

The situation is as follows:

If I run 

```bash
$ quarto render  —execute
[   1/1488] contact.qmd
[   2/1488] post/ismir2019.qmd
[   3/1488] post/regression_intro_2022.qmd
[   4/1488] post/age_equity_dating.qmd
[   5/1488] post/morandini_and_blamey.qmd
[   6/1488] post/masters_thesis.qmd
[   7/1488] post/llm_hacks_2024.qmd
^C⏎
```

it builds the site without executing code. However if I run the following command,  it rebuilds the code in my executable notebooks, pausing at the first one it reaches with real code in.


```bash
$ quarto render  --no-execute
[   1/1488] contact.qmd
[   2/1488] post/ismir2019.qmd
[   3/1488] post/regression_intro_2022.qmd
[   4/1488] post/age_equity_dating.qmd

ERROR: Error executing '/usr/local/bin/Rscript': Broken pipe (os error 32)
R installation:
...
```

In fact it failes in this case, because the test environment is missing some R packages, but you get the idea: it is still trying to run the R code only when the `--no-execute` flag is present. This seems reversed compared to how I imagine this flag working. Am I missing something?

**Reply by cderv** [reply]
Thanks for sharing. It seems that `--no-execute` has some impact in should not have 🤔 
I'll see if I can reproduce. 

Those flag are meant to be used with Jupyter mainly, and I suspect they would be ignore for now with other engines.

**Reply by cderv** [reply]
Ok I found why. I'll answer above following my previous answer: https://github.com/quarto-dev/quarto-cli/discussions/10984#discussioncomment-10867327
